File: src/main/java/svnserver/auth/LocalUserDB.java
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.auth;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import svnserver.UserType;

import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * Simple user db with clear-text passwords.
 *
 * @author Marat Radchenko <marat@slonopotamus.org>
 */
public final class LocalUserDB implements UserDB {

  @NotNull
  private final Map<String, UserWithPassword> users = new HashMap<>();
  @NotNull
  private final Collection<Authenticator> authenticators = Collections.singleton(new CramMD5Authenticator(users::get));

  public LocalUserDB() {
  }

  @Nullable
  public User add(@NotNull String username, @NotNull String password, @NotNull String realName, String email) {
    if (users.containsKey(username)) {
      return null;
    }

    final UserWithPassword userWithPassword = new UserWithPassword(User.create(username, realName, email, username, UserType.Local), password);
    users.put(userWithPassword.getUser().getUsername(), userWithPassword);
    return userWithPassword.getUser();
  }

  @NotNull
  @Override
  public Collection<Authenticator> authenticators() {
    return authenticators;
  }

  @Override
  public User check(@NotNull String username, @NotNull String password) {
    final UserWithPassword userWithPassword = users.get(username);
    if (userWithPassword == null)
      return null;

    if (!userWithPassword.getPassword().equals(password))
      return null;

    return userWithPassword.getUser();
  }

  @Nullable
  @Override
  public User lookupByUserName(@NotNull String username) {
    final UserWithPassword user = users.get(username);
    return user == null ? null : user.getUser();
  }

  @Nullable
  @Override
  public User lookupByExternal(@NotNull String external) {
    return lookupByUserName(external);
  }
}

File: src/main/kotlin/svnserver/auth/LocalUserDB.kt
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.auth

import org.apache.commons.collections4.trie.PatriciaTrie
import svnserver.UserType

/**
 * Simple user db with clear-text passwords.
 *
 * @author Marat Radchenko <marat@slonopotamus.org>
 */
class LocalUserDB : UserDB {
    private val users = PatriciaTrie<UserWithPassword>()

    override val authenticators: Collection<Authenticator> = setOf(CramMD5Authenticator { key: String -> users[key] })

    fun add(username: String, password: String, realName: String, email: String?): User? {
        if (users.containsKey(username)) {
            return null
        }
        val user: User = User.create(username, realName, email, username, UserType.Local, null)
        val userWithPassword = UserWithPassword(user, password)
        users[userWithPassword.user.username] = userWithPassword
        return userWithPassword.user
    }

    override fun check(username: String, password: String): User? {
        val userWithPassword: UserWithPassword = users[username] ?: return null
        if (userWithPassword.password != password) return null
        return userWithPassword.user
    }

    override fun lookupByUserName(username: String): User? {
        return users[username]?.user
    }

    override fun lookupByExternal(external: String): User? {
        return lookupByUserName(external)
    }
}
